niceman/utils.py

```diff
@@ -1255,4 +1256,22 @@ def parse_semantic_version(version):
             "{} does not appear to follow semantic versioning".format(version))
 
 
+# TODO: normalize_command is copied directly from datalad/interface/run.py.
+# It'd be better if the datalad function was moved to its datalad/utils.py, and
+# then we absorb normalize_command through the branch '3rd'.
+
+
+def normalize_command(command):
+    """Convert `command` to the string representation.
+    """
+    if isinstance(command, list):
+        if len(command) == 1:
+            # This is either a quoted compound shell command or a simple
+            # one-item command. Pass it as is.
+            command = command[0]
```

Comment:
what if it is `["executable with space in its name"]`?

Reply:
In the context of `datalad run`, then the user has to the string form:

```
datalad run './stupid\ name.sh'
```

I don't see any reliable alternative.  We'd be left parsing an arbitrary string trying to guess the user's intention.  If you have a suggestion though, I think this is better suited for an issue on datalad.

Reply:
Ok, I will review the assumptions of the datalad run before further comments here...

Reply:
if I got it right, may be here in `niceman` we (for now at least) do not inherit them and always assume that the line is a list of arguments, and if a command as a string is needed, add an explicit cmdline flag for that (e.g. `--cmd-string`)?
May be actually we could add it also in datalad and have it not just bool, but a choice between bool and 'auto', and have 'auto' default which is what it is now deciding between the two modes?  Then whenever it is needed to state explicitly, user could?

Reply:
> Then whenever it is needed to state explicitly, user could?

In my view, that is more complicated and not any better than just saying that the user should just use the string form in this case.

Reply:
And to step back a bit: you're getting bogged down in a known edge case that *has* a solution (use the string form), while this PR fixes multiple issues that are more likely to be encountered and *don't have* a solution.

Reply:
I am "bogged down" to it because that is where with this change I came to realization of this problem exists.  "solution" is not really a solution since it implies that any script (python or shell) needing to run a single command would need always to do escaping of the command (might have spaces in the path etc) even if it provides it explicitly as a list in Python which typically means that no escaping generally is needed at that "user" end  of the story.

Reply:
but ok, let's deal with it later I guess if you feel that it is not "that important" and we have the issue throughout other sessions as well.  you are welcome to merge as is.

Reply:
> it implies that any script (python or shell) needing to run a single command would need always to do escaping of the command (might have spaces in the path etc)

In the context of `datalad run`, the focus is on CLI calls.  In that case, the user needs to quote or otherwise escape "path/with spaces/command.sh" in the same way they would have to in any command line call.  But ...

> even if it provides it explicitly as a list in Python which typically means that no escaping generally is needed at that "user" end of the story.

... thanks for the pushback here.  I'm wrong in trying to apply DataLad's CLI-focused `normalize_command` helper to python methods like `Session.execute_command`.   In the cases in the PR, there's no ambiguity there between list and string [0], so we can just map `shlex_quote` over the items unconditionally.  I'll push an update.

[0] The same thing is true for the rarer form of calling `Dataset.run` rather than `datalad run`.  If there is an easy way of telling that we're being called from the Python API rather than the CLI, we could change how this behavior for `Dataset.run` too.